pkg/updates/memory.go

```diff
@@ -0,0 +1,188 @@
+package updates
+
+import (
+	"fmt"
+
+	"golang.org/x/exp/maps"
+)
+
+// EdgeSet maps a node id to a list of node ids that it can update to
+type EdgeSet map[string][]string
+
+// NodeSet maps a node id to an index in the OrderedNodes array
+type NodeSet map[string]int
+
+// MemorySource is an in-memory implementation of Source.
+// It's an oracle to answer update questions for an installed version.
+type MemorySource struct {
+	// OrderedNodes is an ordered list of all nodes. Lower index == newer version.
+	OrderedNodes []State
+	// Nodes is a helper to lookup a node by id
+	Nodes NodeSet
+	// Edges contains the edgeset for this source.
+	Edges EdgeSet
+}
+
+// Next returns the newest version that can be installed in one step.
+func (m *MemorySource) Next(from string) string {
+	if edges, ok := m.Edges[from]; ok && len(edges) > 0 {
+		return edges[len(edges)-1]
+	}
+	return ""
+}
+
+// NextDirect returns the newest version that can be directly installed without
+// running any migrations.
+func (m *MemorySource) NextDirect(from string) (found string) {
+	initial := m.OrderedNodes[m.Nodes[from]]
+	if to, ok := m.Edges[from]; ok && len(to) > 0 {
+		for _, n := range m.Edges[from] {
+			node := m.OrderedNodes[m.Nodes[n]]
+
+			// if the phase and migration match the current node, no migrations
+			// are required
+			if initial.Phase == node.Phase && initial.Migration == node.Migration {
+				found = n
+			} else {
+				break
+			}
+		}
+	}
+	return found
+}
+
+// Latest returns the newest version that can be installed. If different
+// from `Next`, that means multiple steps are required (i.e. a multi-phase
+// migration, or a required stopping point in a series of updates).
+func (m *MemorySource) Latest(id string) string {
```

Comment:
Maybe we should have a function that returns the versions necessary from one to the other, and if one, then its direct?

Reply:
It's possible to have all three be different: 1.7.0 -> 1.7.1 is `nextdirect`, because no migration is required. 1.7.0->1.9.0 is `next`, because it can be done in one step, but a migration will run. 1.7.0 -> 1.14.1 is `latest` because it's the newest thing that can be reached with some set of steps

Reply:
Hmmm... why treat the "one step but with a migration" differently?